File: mv-fiscal-front/src/app/service/user/user.service.ts
import { Injectable } from '@angular/core';
import { UserDTO } from '../../model/user.dto';
import { Observable } from 'rxjs';
import { ApiService } from '../api/api.service';
import { HttpClient } from '@angular/common/http';
import { environment } from '../../../environment/environment';

@Injectable({
  providedIn: 'root',
})
export class UserService extends ApiService {
  private resourcePath = 'usuario';

  constructor(protected override http: HttpClient) {
    super(http, environment.userApiBaseUrl);
  }

  public getPageByFilter(
    id: number | null,
    name: string,
    email: string,
    createdDate: Date | null,
    pageNumber: number,
    pageSize: number
  ): Observable<any> {
    const params: any = {
      id: id,
      name: name,
      email: email,
      createdDate: createdDate,
      page: pageNumber,
      size: pageSize,
    };

    return this.getWithUrlParams(`${this.resourcePath}`, params);
  }

  public save(userDTO: UserDTO): Observable<any> {
    if (userDTO.id) {
      return this.put(`${this.resourcePath}`, userDTO);
    }
    return this.post(`${this.resourcePath}`, userDTO);
  }

  public deleteById(userId: number) {
    return this.delete(`${this.resourcePath}`, userId);
  }
}

File: mv-fiscal-back/userservice/src/main/java/com/mvfiscal/api/userservice/service/UserService.java
package com.mvfiscal.api.userservice.service;

import com.mvfiscal.api.userservice.dto.SetTaskDTO;
import com.mvfiscal.api.userservice.dto.UserDTO;
import com.mvfiscal.api.userservice.exception.UserDataValidationException;
import com.mvfiscal.api.userservice.model.User;
import com.mvfiscal.api.userservice.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.Date;
import java.util.Optional;

import static java.util.Objects.isNull;
import static java.util.Objects.nonNull;

@Service
@RequiredArgsConstructor
public class UserService {

    private final TaskRestService taskRestService;
    private final UserRepository userRepository;

    public Optional<UserDTO> getUserById(Long userId) {
        return this.userRepository.findUserById(userId);
    }

    public Page<UserDTO> getPageByFilter(Long id, String name, String email, Date createdDate, Pageable pageable) {
        return this.userRepository.findPageByFilter(id, name, email, createdDate, pageable);
    }

    public User saveUser(UserDTO userDTO) {
        return this.userRepository.save(userDTO.asEntity());
    }

    public void validateUserExistence(Long userId) {
        if (!this.userRepository.existsById(userId)) {
            throw new UserDataValidationException(
                    String.format("Nenhum usuário encontrado com o ID: %d", userId));
        }
    }

    private boolean userHasAssociatedTasks(Long userId) {
        return this.taskRestService.userHasAssociatedTasks(userId);
    }

    public User updateUser(UserDTO userDTO) {
        if (isNull(userDTO.getId())) {
            throw new UserDataValidationException("O ID do usuário deve ser informado");
        }
        this.validateUserExistence(userDTO.getId());

        return this.userRepository.save(userDTO.asEntity());
    }

    public void deleteUserById(Long userId) {
        this.validateUserExistence(userId);

        if (this.userHasAssociatedTasks(userId)) {
            throw new UserDataValidationException("O usuário possui tarefas associadas, portanto não pode ser excluído");
        }

        this.userRepository.deleteById(userId);
    }
}
